### PR TITLE
IDEX-2878. Added project type auto-detection for maven and ant projects.

### PR DESCRIPTION
**Pull request by RomanNikitenko**
Added project type auto-detection for maven and ant projects.

### CODE REVIEW

File: platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectService.java [hunk 1]
```diff
@@ -866,7 +870,21 @@ private ImportResponse configureProject(ImportProject importProject, FolderEntry
                 postImportProjectHandler.onProjectImported(project.getBaseFolder());
             }
         } catch (ConflictException | ForbiddenException | ServerException | NotFoundException e) {
-            project = new NotValidProject(baseProjectFolder, projectManager);
+
+            project = projectManager.getProject(workspace, baseProjectFolder.getPath());
+            if (project != null || (!resolveProjectType(baseProjectFolder))) {
```
Comment (sleshchenko) [inline platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectService.java]: I think you can remove a redundant pair of brackets

File: platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectService.java [hunk 2]
```diff
@@ -891,6 +909,16 @@ private ImportResponse configureProject(ImportProject importProject, FolderEntry
         return importResponse;
     }
 
+    private boolean resolveProjectType(FolderEntry projectFolder) {
```
Comment (evoevodin) [inline platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectService.java]: Don't you think that <code>resolveProjectType</code> is not the best name in this case?
Suggestions: <code>isKnownProjectType</code>, <code>hasProjectType</code>.

Also i think that detection of project type is pretty common issue.
Consider moving _detection logic_ to util class and use it! (if it is not done yet).

File: platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/type/ProjectTypeDetectorRegistry.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.Set;
+
+/**
+ * Registry for {@link ProjectTypeDetector}
+ *
+ * @author Roman Nikitenko
+ */
+@Singleton
+public class ProjectTypeDetectorRegistry {
+    private Set<ProjectTypeDetector> detectors = new HashSet<>();
```
Comment (evoevodin) [inline platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/type/ProjectTypeDetectorRegistry.java]: As far as i understand this class must be thread-safe, so __HashSet__ is not the the case, or if you want to use __HashSet__ you must use synchronization for registry methods. Also i think that you should document Thread-safety as well as all methods of this registry, see [example](https://github.com/codenvy/che-core/blob/4.0/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/RuntimeWorkspaceRegistry.java)

File: platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectService.java (hunk 2 above)
Comment (azatsarynnyy) [inline platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectService.java]: I think that we already have the similar functionality there:
ProjectManager#estimateProject()
ProjectManager#resolveSources()

File: platform-api-client-gwt/che-core-client-gwt-project/src/main/java/org/eclipse/che/api/project/gwt/client/ProjectServiceClient.java [hunk 4]
```diff
@@ -104,6 +105,16 @@ void cloneProjectToCurrentWorkspace(String srcWorkspaceId, String srcProjectPath
     void estimateProject(String path, String projectType, AsyncRequestCallback<Map<String, List<String>>> callback);
 
     /**
+     * Resolves sources.
+     *
```
Comment (skabashnyuk) [inline platform-api-client-gwt/che-core-client-gwt-project/src/main/java/org/eclipse/che/api/project/gwt/client/ProjectServiceClient.java]: What does it mean?

**Comment by skabashnyuk** [comment]
ok for me.

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/problemDialog/ProjectProblemDialog.java [hunk 5]
```diff
@@ -33,12 +33,14 @@
      *
      * @param title
      *         the title for popup window
+     * @param keepButTitle
```
Comment (sleshchenko) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/problemDialog/ProjectProblemDialog.java]: mb it better to use keepButtonTitle? "But" is unclear contraction

Comment (skabashnyuk) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/problemDialog/ProjectProblemDialog.java]: +100

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/CoreLocalizationConstant.java [hunk 6]
```diff
@@ -532,8 +532,14 @@
...
 
-    @Key("projectProblem.message")
-    String projectProblemMessage();
+    @Key("projectProblem.blank.message")
+    String projectProblemBlankMessage();
+
+    @Key("projectProblem.estimateTypes.message")
+    String projectProblemEstimateTypesMessage(String types, String keepType);
+
+    @Key("projectProblem.keepBut.title")
```
Comment (sleshchenko) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/CoreLocalizationConstant.java]: The same https://github.com/codenvy/che-core/pull/257#discussion-diff-39620019R36

**Comment by RomanNikitenko** [comment]
The changes have merged in 4ec389521671d1c13f774de4b6458a44a0952354

**Pull request closed without merging**